chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,66 +39,178 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
+
+namespace {
+using FreeBinsIt = std::map<size_t, cuda_internal::FreeList>::iterator;
+}
+
+// Pushes a chunk into an appropriate free list
+//
+// Not thread-safe
+void MemoryPool::PushIntoFreeList(std::unique_ptr<cuda_internal::Chunk> chunk) {
+    cuda_internal::FreeList& free_list = free_bins_[chunk->bytesize()];
+    free_list.emplace_back(std::move(chunk));
+}
+
+void MemoryPool::CompactFreeBins(FreeBinsIt it_start, FreeBinsIt it_end) {
+    auto it_start_rev = std::make_reverse_iterator(it_start);
+    it_start = std::find_if(it_start_rev, free_bins_.rend(), [](const auto& p) { return !p.second.empty(); }).base();
+    it_end = std::find_if(it_end, free_bins_.end(), [](const auto& p) { return !p.second.empty(); });
+    free_bins_.erase(it_start, it_end);
+}
+
+// Finds best-fit, or a smallest larger allocation if available
+//
+// Not thread-safe
+std::unique_ptr<cuda_internal::Chunk> MemoryPool::PopFromFreeList(size_t allocation_size) {
+    FreeBinsIt it_start = free_bins_.lower_bound(allocation_size);
+    FreeBinsIt non_empty_it = std::find_if(it_start, free_bins_.end(), [](const auto& pair) { return !pair.second.empty(); });
+    if (non_empty_it == free_bins_.end()) {
+        return nullptr;
+    }
+    cuda_internal::FreeList& free_list = non_empty_it->second;
+    std::unique_ptr<cuda_internal::Chunk> chunk = std::move(free_list.back());
+    CHAINERX_ASSERT(chunk != nullptr);
+    free_list.pop_back();
+    if (static_cast<size_t>(std::distance(it_start, non_empty_it)) > kCompactionThreashold) {
+        CompactFreeBins(it_start, non_empty_it);
+    }
+    return chunk;
+}
+
+// Removes a chunk from an appropriate free list, and returns the removed chunk
+//
+// Not thread-safe
+std::unique_ptr<cuda_internal::Chunk> MemoryPool::RemoveChunkFromFreeList(cuda_internal::Chunk* chunk) {
+    CHAINERX_ASSERT(chunk != nullptr);
+
+    // Find an appropriate free list
+    FreeBinsIt free_bins_it = free_bins_.find(chunk->bytesize());
+    if (free_bins_it == free_bins_.end()) {
+        return nullptr;
+    }
+    cuda_internal::FreeList& free_list = free_bins_it->second;
+
+    // Remove the given chunk from the found free list
+    auto it = std::find_if(free_list.begin(), free_list.end(), [chunk](const auto& ptr) { return ptr.get() == chunk; });
+    if (it == free_list.end()) {
+        return nullptr;
+    }
+    std::unique_ptr<cuda_internal::Chunk> removed_chunk = std::move(*it);
+    CHAINERX_ASSERT(removed_chunk != nullptr);
+    free_list.erase(it);
+    return removed_chunk;
+}
+
 MemoryPool::~MemoryPool() {
     // NOTE: CudaSetDeviceScope is not available at dtor because it may throw
     int orig_device_index{0};
     cudaGetDevice(&orig_device_index);
     cudaSetDevice(device_index_);
 
-    for (const std::vector<void*>& free_list : free_bins_) {
-        for (void* ptr : free_list) {
-            allocator_->Free(ptr);
+    for (std::pair<const size_t, cuda_internal::FreeList>& pair : free_bins_) {
```

Comment:
It is encouraged to use auto in ranged-for loop https://google.github.io/styleguide/cppguide.html#auto

Reply:
`std::pair<const size_t, cuda_internal::FreeList>` appearants fequently, it might be better to alias this type.

Reply:
It might be another choice, but here we aliases the second with a local variable, so auto does not decrease readability. With this sense, I agree with the google guide (using auto is fine enough). But, I respect your decision.

Reply:
With this sense, using auto is indeed fine enough, but this type appears also in a parameter of lambda expression (currently, written with `auto`), so I want to define an alias of this type and use it in this ranged-for loop.

Reply:
okay

Reply:
Okay, but after reviewing, I felt same sense can be applied for lambda, too.
`pair->second` is not obvious, so we should alias it as a local variable, then `auto& pair` looked fine enough.

Anyway, I respect your decision.